src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php

```diff
@@ -1050,9 +1051,21 @@ protected function isJsonCastable($key)
      */
     protected function isClassCastable($key)
     {
-        return array_key_exists($key, $this->getCasts()) &&
-                class_exists($class = $this->parseCasterClass($this->getCasts()[$key])) &&
-                ! in_array($class, static::$primitiveCastTypes);
+        if (! array_key_exists($key, $this->getCasts())) {
+            return false;
+        }
+
+        $castType = $this->parseCasterClass($this->getCasts()[$key]);
+
+        if (in_array($castType, static::$primitiveCastTypes)) {
+            return false;
+        }
```

Comment:
@Larzs Why are we not allowing primitive cast? This prevent the primary key to be cast

```
$model = new Model();
$model->fill([ 'id' => '123' ]);

// this returns false, because the default primary key `id` is not cast by default, even tho it has a default `int` type 
$model->id === 123;
```

The only way to have this working is to [explicitly providing](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L388) the primary type in the static `$cast` attribute definition of the model, but that sounds like doubling up on the logic `getKeyType()`